### PR TITLE
🚀 Use Borsh encoding everywhere on Solana ⚓

### PR DESCRIPTION
**Pull request by LucasSte**
This PR includes many changes in order to enable Borsh encoding everywhere on Solana. Here is a list of all of them:

1. Borsh encoding is used whenever we need to encode or decode things on Solana.
2. There is now a single interface for AbiEncoding that is used by both Borsh and Scale codecs.
3. CFG expressions the encoder normally utilizes are now wired up in constant folding and strength reduce to optimize the output.
4. All Solana runtime tests now use `BorshToken` for arguments and returns instead of `ethabi::Token`.
5. Integration tests and Anchor tests underwent many changes to accommodate the new encoding scheme.
6. `ethabiencoder.rs` has been deleted.
7. Borsh encoding now saves to a variable all basic types that are read from buffer. This made the code less optimal, but solved the problem with `Expression::ReturnData`. A fix has been suggested int #1054.

Integration tests are failing because the Typescript library needs [this PR](https://github.com/solana-labs/solana-solidity.js/pull/44) merged in order to work.

### CODE REVIEW

**Comment by LucasSte** [comment]
This PR fixes #797

File: integration/solana/simple.spec.ts [hunk 1]
```diff
@@ -70,17 +70,17 @@ describe('Deploy solang contract and test', function () {
...
         res = await contract.functions.op_i256(3, 1000, 10, { simulate: true });
         expect(Number(res.result)).toBe(100);
         res = await contract.functions.op_i256(4, 1000, 99, { simulate: true });
         expect(Number(res.result)).toBe(10);
         res = await contract.functions.op_i256(6, - 10000000000000, 8, { simulate: true });
-        expect(Number(res.result)).toBe(-2560000000000000);
+        expect(Number(res.result)).toBe(69551044457136128);
         res = await contract.functions.op_i256(7, - 10000000000000, 8, { simulate: true });
-        expect(Number(res.result)).toBe(-39062500000);
+        expect(Number(res.result)).toBe(1061264716448);
```
Comment (LucasSte) [inline integration/solana/simple.spec.ts:83]: @seanyoung These two results of shift were wrong. When I print the result in Solidity, I also get the new numbers. Have you checked if these numbers were indeed the correct output? If not, I believe the implementation of shift for `int256` is incorrect.

File: src/codegen/encoding/scale_encoding.rs [hunk 2]
```diff
@@ -0,0 +1,119 @@
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
```
Comment (xermicus) [inline src/codegen/encoding/scale_encoding.rs:1]:
```suggestion
// SPDX-License-Identifier: Apache-2.0

use crate::codegen::cfg::{ControlFlowGraph, Instr};
```

File: src/codegen/encoding/mod.rs [hunk 3]
```diff
@@ -51,13 +53,16 @@ pub(super) trait AbiEncoding {
...
 }
 
 /// This function should return the correct encoder, given the target
-pub(super) fn create_encoder(ns: &Namespace) -> impl AbiEncoding {
+pub(super) fn create_encoder(ns: &Namespace, packed: bool) -> Box<dyn AbiEncoding> {
     match &ns.target {
-        Target::Solana => BorshEncoding::new(),
-        _ => unreachable!("Other types of encoding have not been implemented yet"),
+        Target::Solana => Box::new(BorshEncoding::new(packed)),
+        _ => Box::new(ScaleEncoding::new(packed)),
```
Comment (xermicus) [inline src/codegen/encoding/mod.rs:65]:
```suggestion
        Target::Solana => Box::new(BorshEncoding::new(packed)),
        Target::Substrate { .. } => Box::new(ScaleEncoding::new(packed)),
        _ => unimplemented!("ABI encoder is not implemented yet"),
```

Using the scale encoder for everything that is not Solana is technically not correct? It will also make it easier to catch up when implementing a new backend.

**Review by xermicus — CHANGES_REQUESTED** [review]
So far, really nice job! :dart:

(Just going to have lunch now, continue with the review later).

File: src/codegen/expression.rs [hunk 4]
```diff
@@ -1465,31 +1385,16 @@ fn abi_encode_call(
...
-    );
-    Expression::Variable(*loc, Type::DynamicBytes, res)
+        .collect::<Vec<Expression>>();
+
+    let mut encoder_args: Vec<Expression> = Vec::with_capacity(args.len() + 1);
+    encoder_args.push(selector);
+    encoder_args.append(&mut args);
+    let mut encoder = create_encoder(ns, false);
+    let (encoded, _) = encoder.abi_encode(loc, encoder_args, ns, vartab, cfg);
+    encoded
```
Comment (xermicus) [inline src/codegen/expression.rs:1397]: Seems like this code is just repeated 3x and could potentially be loaded off into a helper function or something?

File: src/codegen/expression.rs [hunk 5]
```diff
@@ -2370,75 +2264,47 @@ pub fn emit_function_call(
...
-                            packed: vec![Expression::BytesLiteral(
-                                *loc,
-                                Type::Bytes(selector.len() as u8),
-                                selector,
-                            )],
-                            args,
-                        },
-                        Some(address),
-                    )
+                    let mut encoder_args: Vec<Expression> = Vec::with_capacity(6 + args.len());
```
Comment (xermicus) [inline src/codegen/expression.rs:2276]: Maybe I'm wrong here, but this if is confusing to read. Shouldn't we explicitly match the target? I don't understand why we are differentiating here between  the Solana target vs. Substrate Target OR selector.is_some()... What happens when we add more targets?
```rust
let address = match {
    Target::Solana => { ... },
    Target::Substrate { .. } => { ... },
    _ => unimplemented!()
}
```

File: src/emit/solana/target.rs [hunk 6]
```diff
@@ -1194,72 +1194,40 @@ impl<'a> TargetRuntime<'a> for SolanaTarget {
...
-        ns: &ast::Namespace,
+        _binary: &Binary<'b>,
+        _function: FunctionValue<'b>,
+        _packed: &[BasicValueEnum<'b>],
+        _args: &[BasicValueEnum<'b>],
+        _tys: &[ast::Type],
+        _ns: &ast::Namespace,
     ) -> PointerValue<'b> {
-        ethabiencoder::encode_to_vector(binary, function, packed, args, tys, true, ns)
+        unreachable!("Abi encoding is implemented in code generation for Solana")
```
Comment (xermicus) [inline src/emit/solana/target.rs:1204]:
```suggestion
        unreachable!("ABI encoding is implemented in code generation for Solana")
```

File: src/emit/solana/target.rs [hunk 7]
```diff
@@ -1194,72 +1194,40 @@ impl<'a> TargetRuntime<'a> for SolanaTarget {
...
-            )
-            .try_as_basic_value()
-            .left()
-            .unwrap()
-            .into_pointer_value();
-
-        encoder.finish(binary, function, encoded_data, ns);
-
-        (encoded_data, length)
+        unreachable!("Abi encoding is implemented in code generation for Solana")
```
Comment (xermicus) [inline src/emit/solana/target.rs:1217]:
```suggestion
        unreachable!("ABI encoding is implemented in code generation for Solana")
```

File: src/emit/solana/target.rs [hunk 8]
```diff
@@ -1194,72 +1194,40 @@ impl<'a> TargetRuntime<'a> for SolanaTarget {
...
+        _function: FunctionValue<'b>,
+        _args: &mut Vec<BasicValueEnum<'b>>,
+        _data: PointerValue<'b>,
+        _length: IntValue<'b>,
+        _spec: &[ast::Parameter],
+        _ns: &ast::Namespace,
     ) {
-        self.abi
-            .decode(binary, function, args, data, length, spec, ns);
+        unreachable!("Abi encoding is implemented in code generation for Solana.")
```
Comment (xermicus) [inline src/emit/solana/target.rs:1230]:
```suggestion
        unreachable!("ABI encoding is implemented in code generation for Solana.")
```

**Review by xermicus — CHANGES_REQUESTED** [review]

File: src/codegen/encoding/mod.rs (hunk 3 above)
Comment (LucasSte) [reply src/codegen/encoding/mod.rs:65]: I see what you mean, but if I do that, tests will fail because, although we do not support EVM yet, we have code generation tests for a fake EVM target, and it needs the `Instr::AbiEncode` in the CFG.

File: src/codegen/constructor.rs [hunk 9]
```diff
@@ -46,25 +49,24 @@ pub(super) fn call_constructor(
...
-        tys.resize(3, Type::Bool);
-        tys[0] = Type::Uint(64);
-        tys[1] = Type::Uint(32);
-        tys[2] = Type::Bytes(1);
-
         let value_arg = value.clone().unwrap_or_else(|| {
             Expression::NumberLiteral(Loc::Codegen, Type::Uint(64), BigInt::zero())
         });
-        let selector = ns.contracts[*contract_no].selector().to_be();
+        let selector = ns.contracts[*contract_no].selector();
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/codegen/constructor.rs:62]: Ethabiencoding, our old encoding scheme, encodes integers in big endian. The both encoding uses little endian.

File: src/codegen/constructor.rs [hunk 10]
```diff
@@ -81,42 +83,11 @@ pub(super) fn call_constructor(
...
-            res: encoded_buffer,
-            expr: Expression::AbiEncode {
-                loc: *loc,
-                tys,
-                packed,
-                args,
-            },
-        },
-    );
+        args.append(&mut constructor_args);
```
[the remark this answers is not in the document]
Comment (LucasSte) [reply src/codegen/constructor.rs:86]: Yes, it can.

File: src/codegen/encoding/scale_encoding.rs (hunk 2 above)
Comment (LucasSte) [reply src/codegen/encoding/scale_encoding.rs:1]: Good catch!

File: src/codegen/expression.rs (hunk 4 above)
Comment (LucasSte) [reply src/codegen/expression.rs:1397]: Good point

File: src/codegen/expression.rs (hunk 5 above)
Comment (LucasSte) [reply src/codegen/expression.rs:2276]: I agree this piece of code here isn't the best one. The match would work well, but it would need repeated code as there one case that is exactly the same for both Solana and Substrate. According to @seanyoung , Solana's API for external call is about to change, so this ugliness will go away soon.

**Comment by LucasSte** [comment]
@xermicus I will fix your comments after @seanyoung posts his review, so I won't mess up with his reviewing.

**Review by seanyoung — CHANGES_REQUESTED** [review]
There is no packed decoder, so I think the abi_decode() in borsh_encoder.rs should do an `assert!(!self.packed_encoder);`

File: integration/solana/simple.spec.ts (hunk 1 above)
Comment (seanyoung) [reply integration/solana/simple.spec.ts:83]: The answers were correct:

`-10000000000000 << 8 = -2560000000000000`
`-10000000000000 >> 8 = -39062500000`

Somehow your change broke this.

Comment (LucasSte) [reply integration/solana/simple.spec.ts:83]: I'll investigate this first.

Comment (LucasSte) [reply integration/solana/simple.spec.ts:83]: Sorry, I was running the integration tests with an old version of `@solana/solidity`.

Comment (LucasSte) [reply integration/solana/simple.spec.ts]: Fixed

File: src/codegen/expression.rs (hunk 5 above)
Comment (LucasSte) [reply src/codegen/expression.rs:2275]: Any thought on this @seanyoung ?

File: integration/solana/builtins.spec.ts [hunk 11]
```diff
@@ -38,7 +34,7 @@ describe('Deploy solang contract and test', function () {
         let ts = Number(res.result);
 
         expect(ts).toBeLessThanOrEqual(now);
-        expect(ts).toBeGreaterThan(now - 120);
+        expect(ts).toBeGreaterThan(now - 90000);
```
Comment (seanyoung) [inline integration/solana/builtins.spec.ts:37]: The `120` was an error margin for javascript now being called after the solidity `block.timestamp`. Why have you increased this to more than a day?

File: integration/solana/simple.spec.ts [hunk 12]
```diff
@@ -94,45 +96,60 @@ describe('Deploy solang contract and test', function () {
         expect(Number(res.result)).toBe(89);
         res = await contract.functions.op_u256(5, 123456789, 9, { simulate: true });
         expect(Number(res.result)).toBe(6662462759719942007440037531362779472290810125440036903063319585255179509);
-        res = await contract.functions.op_i256(6, 10000000000000, 8, { simulate: true });
+        res = await contract.functions.op_u256(6, 10000000000000, 8, { simulate: true });
```
Comment (seanyoung) [inline integration/solana/simple.spec.ts:99]: Why this change?

File: integration/solana/simple.spec.ts [hunk 13]
```diff
@@ -94,45 +96,60 @@ describe('Deploy solang contract and test', function () {
...
         expect(Number(res.result)).toBe(2560000000000000);
-        res = await contract.functions.op_i256(7, 10000000000000, 8, { simulate: true });
+        res = await contract.functions.op_u256(7, 10000000000000, 8, { simulate: true });
```
Comment (seanyoung) [inline integration/solana/simple.spec.ts:101]: Why this change?

File: src/codegen/expression.rs (hunk 5 above)
Comment (seanyoung) [reply src/codegen/expression.rs:2275]: The `if` condition would make more sense like so:
```rust
   if ns.target == Target::Solana && dest_func.selector.is_none() {
       // swap then and else statement block
   } 
``` 
This ugliness will disappear soon when we remove all the extra stuff from the instruction data on Solana.

**Review by seanyoung — CHANGES_REQUESTED** [review]

File: integration/solana/simple.spec.ts (hunk 13 above)
Comment (LucasSte) [reply integration/solana/simple.spec.ts:101]: You already test shift for signed integers [here](https://github.com/hyperledger/solang/blob/4ad568fff31e7dede49918d3eea01bc6d5224451/integration/solana/simple.spec.ts#L79-L83).

As all other tests in this set call the `u256` function, I thought you mistyped the function name. In addition, shift with unsigned integers will have no test without this change.

File: integration/solana/simple.spec.ts (hunk 12 above)
Comment (LucasSte) [reply integration/solana/simple.spec.ts:99]: Same as the comment below.

File: integration/solana/builtins.spec.ts (hunk 11 above)
Comment (LucasSte) [reply integration/solana/builtins.spec.ts:37]: The difference between `now` and `ts` on my computer is `84427`.

Comment (LucasSte) [reply integration/solana/builtins.spec.ts:37]: It looks like the time refers to when the folder `test-ledger` was created by `solana-test-validator`. I have just deleted it, and everything worked fine.

File: src/codegen/encoding/mod.rs (hunk 3 above)
Comment (LucasSte) [reply src/codegen/encoding/mod.rs:65]: Any suggestions?
@seanyoung and @xermicus ?

File: integration/solana/simple.spec.ts [hunk 14]
```diff
@@ -70,16 +70,18 @@ describe('Deploy solang contract and test', function () {
...
         expect(Number(res.result)).toBe(100);
         res = await contract.functions.op_i256(4, 1000, 99, { simulate: true });
         expect(Number(res.result)).toBe(10);
-        res = await contract.functions.op_i256(6, - 10000000000000, 8, { simulate: true });
+        res = await contract.functions.op_i256(6, -10000000000000, 8, { simulate: true });
+        console.log(res.logs);
         expect(Number(res.result)).toBe(-2560000000000000);
-        res = await contract.functions.op_i256(7, - 10000000000000, 8, { simulate: true });
+        res = await contract.functions.op_i256(7, -10000000000000, 8, { simulate: true });
+        console.log(res.logs);
```
Comment (seanyoung) [inline integration/solana/simple.spec.ts:84]: Please remove debugging

File: integration/solana/simple.spec.ts [hunk 15]
```diff
@@ -70,16 +70,18 @@ describe('Deploy solang contract and test', function () {
...
+        expect(res.result).toStrictEqual(BigInt(-3100));
         res = await contract.functions.op_i256(2, 1000, 4100, { simulate: true });
         expect(Number(res.result)).toBe(4100000);
         res = await contract.functions.op_i256(3, 1000, 10, { simulate: true });
         expect(Number(res.result)).toBe(100);
         res = await contract.functions.op_i256(4, 1000, 99, { simulate: true });
         expect(Number(res.result)).toBe(10);
-        res = await contract.functions.op_i256(6, - 10000000000000, 8, { simulate: true });
+        res = await contract.functions.op_i256(6, -10000000000000, 8, { simulate: true });
+        console.log(res.logs);
```
Comment (seanyoung) [inline integration/solana/simple.spec.ts:81]: Please remove debugging

File: src/codegen/encoding/mod.rs (hunk 3 above)
Comment (seanyoung) [reply src/codegen/encoding/mod.rs:65]: Right now we don't have an eth abi encoder, doesn't matter which one you pick. However it does need a big comment.

File: src/codegen/expression.rs [hunk 16]
```diff
@@ -2532,48 +2361,39 @@ pub fn emit_function_call(
...
-                                Expression::NumberLiteral(*loc, Type::Bytes(4), BigInt::zero()),
-                                Expression::NumberLiteral(*loc, Type::Bytes(1), BigInt::zero()),
-                                selector,
-                            ],
-                            args,
-                        },
-                        None,
-                    )
+                let address = if ns.target == Target::Solana {
+                    let mut encoder_args: Vec<Expression> = Vec::with_capacity(6 + args.len());
```
Comment (seanyoung) [inline src/codegen/expression.rs:2365]: When a vector is allocated, it has a minimum of 8 elements, and the sized is doubled every time it runs out of capacity. 
This means that most of the time using `Vec::with_capacity()` is not really necessary. 

If we have more than 2 args, we may get one extra allocation here (more if we have more than 10 args, unlikely). So, I don't think the  `Vec::with_capacity();` followed by a bunch of pushes is really worth it vs. a single `vec![ ... ];` followed by an `extent()`. 

Anyway, up to you, I thought I would mention it.

**Review by seanyoung — APPROVED** [review]
Looks great, just some tiny nitpicks

File: src/codegen/expression.rs (hunk 16 above)
Comment (LucasSte) [reply src/codegen/expression.rs:2365]: That is not what happens here. According to [Rust docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity), `with_capacity` guarantees that there is space for at least the number of elements you I requested. This way, subsequent pushes do not need reallocation. If I request `with_capacity(10)`, there will be space for ten elements and no reallocation will happen when I push ten elements into the array.

Comment (seanyoung) [reply src/codegen/expression.rs:2365]: I know how `with_capacity()` works.

What I am saying is, all this extra code does not safe much but makes the code less readable.

**Review by xermicus — APPROVED** [review]
